### PR TITLE
refactor wallet extension to use a map of clients

### PR DESCRIPTION
**Pull request by BedrockSquirrel**
### Why is this change needed?

- Pre-requisite for the go-obs client
- (hopefully) makes the RPC client logic layers easier to reason about
- The EncRPCClient introduced here is used in *both* the wallet extension and the AuthClient that's coming.

This change also moves the multi-acc fuzzy stuff out of the core client and just into wallet extension now for cleanliness.

### What changes were made as part of this PR:

- Add viewing key type per Tudor's suggestion to wrap account with viewing key and signature
- Rename ViewingKeyClient -> EncRPCClient, a VK-encrypting/decrypting client which is bound to a single account for its lifespan
- rework wallet extension to use a map of acc addresses -> EncRPCClients, with logic to pick the right one or retry them all

Note: this change means we will round-robin all registered clients when it's not obvious which to use, we should monitor and optimise this over time.

Note2: as Joel pointed out this means we find and cast the params map twice (in wallet ext to deduce the correct client, in EncRPCClient to make sure the 'from' is set)

### What are the key areas to look at

EncRPCClient, WalletExtension

### :rotating_light: Definition of Done :rotating_light:
- [ ] Merged into dev-testnet 
- [ ] Passes tests on dev-testnet
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: go/rpcclientlib/encrypted_client.go [hunk 1]
```diff
@@ -4,19 +4,15 @@ import (
...
 	"github.com/ethereum/go-ethereum/crypto/ecies"
 )
 
 const (
-	http                = "http://"
-	reqJSONKeyFrom      = "from"
-	reqJSONKeyData      = "data"
-	ethCallPaddedArgLen = 64
-	ethCallAddrPadding  = "000000000000000000000000"
+	http           = "http://"
```
Comment (tudor-malene) [inline go/rpcclientlib/encrypted_client.go:14]: This file needs to have a single responsibility, which is the encClient.
All the other stuff should be removed.
It should start with the `EncRPCClient` struct, and the comment that it's a wrapper..

For example these will not be needed after the suggested refactorings.

**Review by tudor-malene — CHANGES_REQUESTED** [review]
Thanks Matt. I think this is definitely much clearer. 
I left a few suggestions, which should clarify and simplify further.

File: go/rpcclientlib/encrypted_client.go [hunk 2]
```diff
@@ -216,7 +209,7 @@ func (c *ViewingKeyClient) RegisterViewingKey(signature []byte, signerAddress co
 // The enclave requires the `from` field to be set so that it can encrypt the response, but sources like MetaMask often
 // don't set it. So we check whether it's present; if absent, we walk through the arguments in the request's `data`
 // field, and if any of the arguments match our viewing key address, we set the `from` field to that address.
-func (c *ViewingKeyClient) setFromFieldIfMissing(args []interface{}) ([]interface{}, error) {
+func (c *EncRPCClient) setFromFieldIfMissing(args []interface{}) ([]interface{}, error) {
```
[the remark this answers is not in the document]
Comment (BedrockSquirrel) [reply go/rpcclientlib/encrypted_client.go:212]: This does set it to the address of the current viewing key. But perhaps it shouldn't check (so having from set is required) and the wallet extension should make sure it sets it if needed before each attempted request

[the remark this answers is not in the document]
Comment (tudor-malene) [reply go/rpcclientlib/encrypted_client.go:212]: ah, ok. then the comment is wrong.
I'd say we add a check that the from field, if set, is the current vk.account. If not set, we set it to that

File: tools/walletextension/wallet_extension.go [hunk 3]
```diff
@@ -162,8 +160,25 @@ func (we *WalletExtension) handleHTTPEthJSON(resp http.ResponseWriter, req *http
 		return
 	}
 
+	// for obscuro RPC requests it is important we know the sender account for the viewing key encryption/decryption
```
Comment (tudor-malene) [inline tools/walletextension/wallet_extension.go:163]: feels like all this logic from 163 to 181 belong in a function in the multi_acc_helper

**Review by tudor-malene — COMMENTED** [review]
looks good. One small thing

File: tools/walletextension/multi_acc_helper.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	var err error
+	if suggestedClient != nil {
+		// todo: if we have a suggested client, should we still loop through the other clients if it fails?
+		// 		The call data guessing won't often be wrong but there could be edge-cases there
+		err = executeCall(suggestedClient, rpcReq, &rpcResp)
+	} else {
+		// we attempt the request with every client until we have a successful execution
+		log.Info("appropriate client not found, attempting request with up to %d clients", len(accClients))
+		for _, client := range accClients {
+			err = executeCall(client, rpcReq, &rpcResp)
```
Comment (tudor-malene) [inline tools/walletextension/multi_acc_helper.go:160]: one last thing.
executeCall should prob be in the same file.

**Review by tudor-malene — APPROVED** [review]
lgtm.
one minor function move, and good to merge.

File: tools/walletextension/multi_acc_helper.go (hunk 4 above)
Comment (BedrockSquirrel) [reply tools/walletextension/multi_acc_helper.go:160]: Oops, yeah good spot :)

**Pull request merged**